jni/src/Foreign/JNI/Types.hs

```diff
@@ -220,7 +227,7 @@ instance Storable JValue where
   poke p (JLong x) = poke (castPtr p) x
   poke p (JFloat x) = poke (castPtr p) x
   poke p (JDouble x) = poke (castPtr p) x
-  poke p (JObject x) = poke (castPtr p :: Ptr (J a)) x
+  poke p (JObject (J x)) = withForeignPtr x (poke (castPtr p))
```

Comment:
I think we better use our own version of `withArray` that can handle this by calling `withForeignPtr` as needed. It will be safe and won't need to create new references.